ICRC-29 Allow redirect by signer to different origin (#182)
dfinity/wg-identity-authentication: topics/icrc_29_window_post_message_transport.md

}
```

The signer should reply with a message indicating that it is ready to receive additional messages.

> The message `targetOrigin` should be set to the previously received `icrc29_status` message its `origin` property value.

Reviewer: This seems a bit difficult to read. How about:

> The message `targetOrigin` should be set to the `origin` value of the received message `icrc29_status`.

using the [window.postMessage](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage) API.

## Signer

The `origin` and `source` values of the received message `icrc29_status` when the communication channel was established,
are mentioned below as `establishedOrigin` and `establishedSource` respectively.

Messages are received by listening to `message` events, messages are considered as coming from relying party if both:
- The received message `origin` property is equal to the `establishedOrigin`.
- The received message `source` property is equal to the `establishedSource`.

Messages are sent by calling the `postMessage` method on the `establishedSource` with the `targetOrigin` parameter set to `establishedOrigin`.

The window must not be automatically closed after sending a message, since the relying party could possibly send additional messages. 
Instead, the relying party is responsible for closing the signer window.

## Relying party

> The `origin` value of the received message `"result": "ready"` when the communication channel was established,

Reviewer 2: ```suggestion
The `origin` value of the first received reply to a `icrc29_status` message with `"result": "ready"` when establishing the communication channel,
```

is mentioned below as `establishedOrigin`.

The `window` that was opened for the signer is mentioned below as `signerWindow`.

> Messages are received by listening to `message` events, messages are considered as coming from signer if both:

Reviewer 2: ```suggestion
Messages are received by listening to `message` events. Messages are considered as coming from signer if both:
```

- The received message `origin` property is equal to the `establishedOrigin`.
- The received message `source` property is equal to the `signerWindow`.

Messages are sent by calling the `postMessage` method on the `signerWindow` with the `targetOrigin` parameter set to `establishedOrigin`.

> Make sure to call the `close` method on the `signerWindow` after it is no longer needed.  

Reviewer 2: ```suggestion
The relying party should call the `close` method on the `signerWindow` after it is no longer needed.  
```

> After having closed the window, the signer must again go through the process of [establishing a communication channel](#establishing-a-communication-channel) again.

Reviewer 2: ```suggestion
After having closed the window, the signer must again go through the process of [establishing a communication channel](#establishing-a-communication-channel).
```

> **Note**: The relying party should send `icrc29_status` messages in short intervals. It is expected that some of the messages will be lost due to being sent before the signer is ready.

After the `"result": "ready"` response has been received within a reasonable timeframe, the relying party can send [ICRC-25](https://github.com/dfinity/wg-identity-authentication/blob/main/topics/icrc_25_signer_interaction_standard.md) messages to the signer
using the [window.postMessage](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage) API.

## Authentication

* The relying party must authenticate the signer by checking if both:

>   - The received message `source` property is equal to the `window` that was opened.

Reviewer 2: Doesn't this defeat the purpose of redirecting? I.e. after a redirect, this will never match?

Author: As discussed in WG today, I wasn't aware that a message eventsource could be used instead of window.opener. I'll update this MR.

  - The received message `origin` property is equal to the `origin` property in the `"result": "ready"` response received when communication channel was established.
* The signer must authenticate the relying party by checking if both:
  - The received message `source` property is equal to the `source` property in the `icrc29_status` message received when communication channel was established.
  - The received message `origin` property is equal to the `origin` property in the `icrc29_status` message received when communication channel was established.

## Sending Messages

> Messages are sent by calling `window.postMessage` on the signer `window`, or the `source` property in the `icrc29_status` message received when communication channel was established respectively.

Reviewer: How about having two subsections within `Sending Messages`. One for the relying party and another for the signer.

Or make it two sections: `Sending Messages: Signer` and `Sending Messages: Reying Party`.

The recommendations here are intertwined and a dev building one or the other should have it easy to find how to deal with this. The same way that the `Authentication` section has two points.

Author: Good points, it was indeed getting harder to read 😅
